File: android/app/src/main/java/com/videoconverter/android/domain/Naming.kt
package com.videoconverter.android.domain

fun sourceStem(displayName: String): String {
    val name = displayName.substringAfterLast('/', displayName)
    val stem = name.substringBeforeLast('.', name)
    return stem.ifEmpty { "output" }
}

fun partialOutputPath(output: String): String {
    val lastSlash = output.lastIndexOf('/')
    val parent = if (lastSlash >= 0) output.substring(0, lastSlash) else ""
    val fileName = if (lastSlash >= 0) output.substring(lastSlash + 1) else output

    val dotIndex = fileName.lastIndexOf('.')
    val ext = if (dotIndex >= 0) fileName.substring(dotIndex + 1) else "bin"
    val stem = if (dotIndex >= 0) fileName.substring(0, dotIndex) else fileName
    val actualStem = stem.ifEmpty { "output" }

    val partialName = "$actualStem.partial.$ext"
    return if (parent.isNotEmpty()) "$parent/$partialName" else partialName
}

fun ffmpegFileArg(path: String): String {
    if (path.startsWith("file:") || path.startsWith("pipe:") || path.startsWith("/proc/self/fd/")) {
        return path
    }
    return "file:$path"
}

fun uniqueFileName(
    stem: String,
    ext: String,
    taken: (String) -> Boolean,
    clock: () -> String = ::outputCollisionStamp,
): String {
    val base = "$stem.$ext"
    if (!taken(base)) return base
    val stamp = clock()
    val stamped = "${stem}_$stamp.$ext"
    if (!taken(stamped)) return stamped
    var index = 1
    while (taken("${stem}_${stamp}_$index.$ext")) {
        index++
    }
    return "${stem}_${stamp}_$index.$ext"
}

fun outputCollisionStamp(): String {
    val now = java.time.LocalDateTime.now()
    return java.time.format.DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss").format(now)
}

fun allocateOutputPath(
    outputDir: String,
    stem: String,
    ext: String,
    exists: (String) -> Boolean,
    clock: () -> String = ::outputCollisionStamp,
): String {
    val dir = outputDir.trimEnd('/')
    val name = uniqueFileName(stem, ext, { exists("$dir/$it") }, clock)
    return "$dir/$name"
}

fun sanitizeRenameStem(raw: String): String? {
    val trimmed = raw.trim()
    if (trimmed.isEmpty() || trimmed.contains('/') || trimmed.contains('\\')) return null
    val cleaned = trimmed
        .replace(Regex("""[:*?"<>|]"""), "")
        .replace(Regex("""\s+"""), " ")
        .trim()
        .trim('.')
    val stem = cleaned.substringBeforeLast('.', cleaned).trim()
    if (stem.isEmpty() || stem == "." || stem == "..") return null
    return stem.take(80)
}

fun renamedFileName(displayName: String, outputPath: String?, stem: String): String {
    val current = outputPath
        ?.substringAfterLast('/')
        ?.substringAfterLast('\\')
        ?.substringBefore('?')
        ?.takeIf { it.isNotBlank() }
        ?: displayName
    val ext = current.substringAfterLast('.', "").takeIf { it.isNotBlank() && !it.contains('/') }
        ?: displayName.substringAfterLast('.', "").takeIf { it.isNotBlank() }
        ?: ""
    return if (ext.isBlank()) stem else "$stem.$ext"
}

fun canRenameJob(status: JobStatus): Boolean = status == JobStatus.Completed

File: ios/LiteTrans/Domain/Naming.swift
import Foundation

public func sourceStem(_ displayName: String) -> String {
    let name = displayName.split(separator: "/").last.map(String.init) ?? displayName
    if let dot = name.lastIndex(of: "."), dot != name.startIndex {
        let stem = String(name[..<dot])
        return stem.isEmpty ? "output" : stem
    }
    return name.isEmpty ? "output" : name
}

public func partialOutputPath(_ output: String) -> String {
    let lastSlash = output.lastIndex(of: "/")
    let parent = lastSlash.map { String(output[..<$0]) } ?? ""
    let fileName = lastSlash.map { String(output[output.index(after: $0)...]) } ?? output
    let ext: String
    let stem: String
    if let dot = fileName.lastIndex(of: ".") {
        ext = String(fileName[fileName.index(after: dot)...])
        stem = String(fileName[..<dot])
    } else {
        ext = "bin"
        stem = fileName
    }
    let actualStem = stem.isEmpty ? "output" : stem
    let partialName = "\(actualStem).partial.\(ext)"
    return parent.isEmpty ? partialName : "\(parent)/\(partialName)"
}

public func allocateOutputPath(outputDir: String, stem: String, ext: String, exists: (String) -> Bool) -> String {
    let dir = outputDir.hasSuffix("/") ? String(outputDir.dropLast()) : outputDir
    let candidate = "\(dir)/\(stem).\(ext)"
    if !exists(candidate) { return candidate }
    var index = 1
    while true {
        let numbered = "\(dir)/\(stem)-\(index).\(ext)"
        if !exists(numbered) { return numbered }
        index += 1
    }
}

public func sanitizeRenameStem(_ raw: String) -> String? {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty || trimmed.contains("/") || trimmed.contains("\\") { return nil }
    var cleaned = trimmed
    for ch in [":", "*", "?", "\"", "<", ">", "|"] {
        cleaned = cleaned.replacingOccurrences(of: ch, with: "")
    }
    cleaned = cleaned.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
        .trimmingCharacters(in: .whitespacesAndNewlines)
        .trimmingCharacters(in: CharacterSet(charactersIn: "."))
    let stem = cleaned.split(separator: ".").dropLast().joined(separator: ".")
    let resolved = cleaned.contains(".") ? stem : cleaned
    if resolved.isEmpty || resolved == "." || resolved == ".." { return nil }
    return String(resolved.prefix(80))
}

public func canRenameJob(_ status: JobStatus) -> Bool { status == .completed }

public func numberedOutputName(stem: String, index: Int, ext: String) -> String {
    String(format: "%@-%03d.%@", stem, index, ext)
}

public func ffmpegFileArg(_ path: String) -> String {
    if path.hasPrefix("file:") || path.hasPrefix("pipe:") { return path }
    return "file:\(path)"
}

public func photosImportDisplayName(
    fileName: String,
    pathExtension: String = "",
    itemIdentifier: String? = nil
) -> String {
    if isUsefulImportDisplayName(fileName) {
        return fileName
    }
    if let itemIdentifier, isUsefulImportDisplayName(itemIdentifier) {
        return itemIdentifier
    }
    let ext = normalizedImportExtension(pathExtension.isEmpty ? fileExtension(of: fileName) : pathExtension)
    switch ext {
    case "mp4": return "video.mp4"
    case "mov": return "video.mov"
    case "": return "video.mov"
    default: return "video.\(ext)"
    }
}

func isUsefulImportDisplayName(_ name: String) -> Bool {
    let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return false }
    let firstComponent = trimmed.split(separator: "/").first.map(String.init) ?? trimmed
    let stem = sourceStem(firstComponent)
    return UUID(uuidString: stem) == nil && UUID(uuidString: firstComponent) == nil
}

private func fileExtension(of name: String) -> String {
    guard let dot = name.lastIndex(of: "."), dot != name.startIndex else { return "" }
    return String(name[name.index(after: dot)...])
}

private func normalizedImportExtension(_ ext: String) -> String {
    ext.trimmingCharacters(in: CharacterSet(charactersIn: ".")).lowercased()
}
